languages/cpp/reference/README.md

```diff
@@ -0,0 +1,158 @@
+# C++ reference
+
+## Concepts
+
+The C++ concept exercises are based on concepts. The list below contains the concepts that have been identified for the C++ language.
+
+### Object-oriented
+
+The core OO features a C++ developer should know about are:
+
+- Initialization
+- [Encapsulation](../../../reference/concepts/encapsulation.md)
+- [Classes](../../../reference/concepts/classes.md)
+- [Objects](../../../reference/concepts/objects.md)
+- [State](../../../reference/concepts/state.md)
+- [Mutation](../../../reference/concepts/mutation.md)
+- [Composition](../../../reference/concepts/composition.md)
+- [Inheritance](../../../reference/concepts/inheritance.md)
+- [Interfaces](../../../reference/concepts/interfaces.md)
+- [Runtime polymorphism](../../../reference/concepts/polymorphism.md)
+- [Methods](../../../reference/concepts/methods.md)
+
+### Functional
+
+C++ is a very flexible language, so lots of functional concepts apply to it:
+
+- [Immutability](../../../reference/concepts/immutability.md) (`const`ness)
+- [Higher-order functions](../../../reference/concepts/higher_order_functions.md)
+- [Type inference](../../../reference/concepts/type_inference.md) (`auto`, FTAD, and CTAD)
+- [Anonymous functions (lambdas)](../../../reference/concepts/anonymous_functions.md)
+- [Recursion](../../../reference/concepts/recursion.md)
+- [Nested functions](../../../reference/concepts/nested_functions.md)
+- [Pipelines (ranges)](../../../reference/concepts/pipelines.md) (Defer C++2a to release) 
+
+### Memory management
+
+- Object lifetime
+- RAII
+- Move-semantics
```

Comment:
In Memory management I would add:
* Stack & Heap
* Raw pointers
* Smart pointers

Reply:
I thought of classifying smart-pointers in RAII
Stack and heap... well, technically the standard doesn't define stack at all; it does define the concept of a "free-store" however, so I'll add it

Reply:
They're a pretty important way to manage memory in a safe manner in modern C++. I think a student should have a good grasp of why they exist and how they can be used, joining them with RAII would be a bit too diminuitive I think.

Reply:
Keep in mind that this is concepts; exercises within a given concept are more specialized

Reply:
Yeah, right. I didn't think about that.